src/monodroid/jni/monodroid-glue.cc

```diff
@@ -1125,14 +1121,77 @@ MonodroidRuntime::monodroid_dlopen_log_and_return (void *handle, char **err, con
 		delete[] full_name;
 	}
 
+#if !defined (NET6)
 	if (!need_api_init)
 		return handle;
-
 	init_internal_api_dso (handle);
+#endif
 
 	return handle;
 }
 
+#if defined (NET6) && defined (ANDROID)
+force_inline void*
+MonodroidRuntime::monodroid_dlopen (const char *name, int flags, char **err)
+{
+	unsigned int dl_flags = monodroidRuntime.convert_dl_flags (flags);
+	void *h = androidSystem.load_dso_from_any_directories (name, dl_flags);
+	if (h != nullptr) {
+		return monodroid_dlopen_log_and_return (h, err, name, false /* name_needs_free */);
+	}
+
+	if (!utils.ends_with (name, ".dll.so")) {
+		return monodroid_dlopen_log_and_return (h, err, name, false /* name_needs_free */);
+	}
+
+	char *basename_part = const_cast<char*> (strrchr (name, '/'));
+	if (basename_part != nullptr) {
+		basename_part++;
+	} else {
+		basename_part = (char*)name;
+	}
+
+	constexpr char LIBAOT_PREFIX[] = "libaot-";
+	constexpr size_t LIBAOT_PREFIX_SIZE = sizeof(LIBAOT_PREFIX);
+
+	size_t base_len = strlen (basename_part);
+	size_t len = base_len + LIBAOT_PREFIX_SIZE; // includes the trailing \0
+	static_local_string<PATH_MAX> basename (len);
+
+	basename.append (LIBAOT_PREFIX).append (basename_part);
+
+	h = androidSystem.load_dso_from_any_directories (basename.get (), dl_flags);
+
+	if (h != nullptr && XA_UNLIKELY (utils.should_log (LOG_ASSEMBLY))) {
+		log_info_nocheck (LOG_ASSEMBLY, "Loaded AOT image '%s'", basename.get ());
+	}
+
+	return h;
+}
+
+void*
+MonodroidRuntime::monodroid_dlopen (const char *name, int flags, char **err, [[maybe_unused]] void *user_data)
+{
+	log_warn (LOG_DEFAULT, "MonodroidRuntime::monodroid_dlopen (\"%s\", 0x%x, %p, %p)", name, flags, err, user_data);
+	if (name == nullptr) {
+		log_warn (LOG_ASSEMBLY, "monodroid_dlopen got a null name. This is not supported in NET6+");
+		return nullptr;
+	}
+
+	constexpr char DSO_EXTENSION[] = ".so";
+	constexpr size_t DSO_EXTENSION_SIZE = sizeof(DSO_EXTENSION);
+
+	if (utils.ends_with (name, DSO_EXTENSION)) {
+		return monodroid_dlopen (name, flags, err);
+	}
+
+	size_t len = strlen (name) + DSO_EXTENSION_SIZE; // includes the trailing \0
```

Comment:
Do we check/how do we prevent. `strlen(name)` from being `std::numeric_limits<size_t>::max`, causing the addition to be an overflow?

Should this use one of our "checked addition" macros?